### PR TITLE
W3DProjectedShadow

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

**Comment by tomsons26** [comment]
Add_ShadowDrawCalls > Add_Draw_Calls and the member it sets, its general purpose

**Comment by tomsons26** [comment]
Unk call Gather_Draw_Stats, it takes the DebugDrawStats class

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 1]
```diff
@@ -0,0 +1,1581 @@
...
+
+int g_shadowDecalVertexSize = 32768;
+int g_shadowDecalIndexSize = 65536;
+
+#ifndef GAME_DLL
+W3DProjectedShadowManager *g_theW3DProjectedShadowManager = nullptr;
+FrustumClass *g_shadowCameraFrustum;
+#endif
+
+W3DShadowTexture::W3DShadowTexture()
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: init name and texture, comment as bugfix

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 2]
```diff
@@ -0,0 +1,1581 @@
...
+                return 0;
+            }
+        }
+        tex->Release_Ref();
+    }
+    return 1;
+}
+
+W3DProjectedShadow::W3DProjectedShadow()
+{
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp:267]: init robj, next, uvs, comment as bugfix

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 3]
```diff
@@ -0,0 +1,1581 @@
...
+W3DProjectedShadow::~W3DProjectedShadow()
+{
+    Ref_Ptr_Release(m_shadowProjector);
+
+    for (int i = 0; i < 1; i++) {
+        Ref_Ptr_Release(m_shadowTexture[i]);
+    }
+}
+
+W3DProjectedShadowManager::W3DProjectedShadowManager() :
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp:408]: init render target members, comment as bugfix

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 4]
```diff
@@ -0,0 +1,1581 @@
...
+            D3DPOOL_DEFAULT,
+            &g_shadowDecalIndexBufferD3D)
+        < S_OK) {
+        return false;
+    }
+
+    return g_shadowDecalVertexBufferD3D
+        || dev->CreateVertexBuffer(24 * g_shadowDecalVertexSize,
+               D3DUSAGE_WRITEONLY | D3DUSAGE_DYNAMIC,
+               0,
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp:492]: D3DFMT_UNKNOWN?

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 5]
```diff
@@ -0,0 +1,1581 @@
...
+    if (dev->CreateIndexBuffer(2 * g_shadowDecalIndexSize,
+            D3DUSAGE_WRITEONLY | D3DUSAGE_DYNAMIC,
+            D3DFMT_INDEX16,
+            D3DPOOL_DEFAULT,
+            &g_shadowDecalIndexBufferD3D)
+        < S_OK) {
+        return false;
+    }
+
+    return g_shadowDecalVertexBufferD3D
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp:460]: flip this so its like above

File: src/game/client/shadow/w3dprojectedshadow.h [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    int Render_Projected_Terrain_Shadow(W3DProjectedShadow *shadow, AABoxClass &box);
+
+    W3DProjectedShadow *Create_Decal_Shadow(Shadow::ShadowTypeInfo *shadow_info);
+    void Queue_Decal(W3DProjectedShadow *shadow);
+    void Flush_Decals(W3DShadowTexture *texture, ShadowType type);
+
+    void Invalidate_Cached_Light_Positions();
+    void Update_Render_Target_Textures();
+
+    SpecialRenderInfoClass *Get_Shadow_Context() { return m_shadowContext; }
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.h]: Get_Render_Context

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 7]
```diff
@@ -0,0 +1,1581 @@
...
+    shadow->m_decalOffsetV = y;
+    shadow->m_offsetX = 0.0f;
+    shadow->m_offsetY = 0.0f;
+    shadow->m_flags = (type & SHADOW_DIRECTIONAL_PROJECTION) != 0;
+    shadow->Init();
+    W3DProjectedShadow *s = nullptr;
+    W3DProjectedShadow *i;
+
+    for (i = m_decalList; i; i = i->m_next) {
+        if (i->m_shadowTexture[0] = tex) {
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: ==

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 8]
```diff
@@ -0,0 +1,1581 @@
...
+    shadow->m_decalOffsetV = y;
+    shadow->m_offsetX = offsetx;
+    shadow->m_offsetY = offsety;
+    shadow->m_flags = flags;
+    shadow->Init();
+    W3DProjectedShadow *s = nullptr;
+    W3DProjectedShadow *i;
+
+    for (i = m_decalList; i; i = i->m_next) {
+        if (i->m_shadowTexture[0] = tex) {
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: ==

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 9]
```diff
@@ -0,0 +1,1581 @@
...
+    shadow->m_decalOffsetV = 1.0f / y;
+    shadow->m_offsetX = offsetx;
+    shadow->m_offsetY = offsety;
+    shadow->m_flags = flags;
+    shadow->Init();
+    W3DProjectedShadow *s = nullptr;
+    W3DProjectedShadow *i;
+
+    for (i = m_decalList; i; i = i->m_next) {
+        if (i->m_shadowTexture[0] = tex) {
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: ==

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 10]
```diff
@@ -0,0 +1,1581 @@
...
+                }
+
+                tex = new W3DShadowTexture();
+                tex->Set_Name(fname);
+                m_W3DShadowTextureManager->Add_Texture(tex);
+                tex->Set_Texture(t);
+            }
+
+            type = SHADOW_DECAL;
+            int flags = (type & SHADOW_DIRECTIONAL_PROJECTION) != 0;
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: shouldn't have int, float, float, ends up never being used

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 11]
```diff
@@ -0,0 +1,1581 @@
...
+                        return;
+                    }
+
+                    g_nShadowDecalStartBatchVertex = 0;
+                    g_nShadowDecalPolysInBatch = 0;
+                    g_nShadowDecalVertsInBatch = 0;
+                    g_nShadowDecalVertsInBuf = 0;
+                }
+
+                captainslog_dbgassert(i11 == (i2 - i1 + 1) * (i4 - i3 + 1), "queueDecal VB size mismatch");
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: BFME2 shows what the names are, `numVerts == ((endY - startY + 1)*(endX - startX + 1))`
seems in binary its `numVerts != (endX - startX + 1) * (endY - startY + 1) `

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 12]
```diff
@@ -0,0 +1,1581 @@
...
+                i3 += i8;
+                i4 -= i7 - i8;
+            }
+
+            int i9 = i2 - i1 + 1;
+            int i10 = i4 - i3 + 1;
+
+            if (i9 > 1 && i10 > 1) {
+                int i11 = i10 * i9;
+                int i12 = 6 * (i4 - i3) * (i2 - i1);
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: i12 is numIndex

File: src/game/client/shadow/w3dshadow.h [hunk 13]
```diff
@@ -13,17 +13,73 @@
...
+    SHADOW_DIRECTIONAL_PROJECTION = 16,
+    SHADOW_ALPHA_DECAL = 32,
+    SHADOW_ADDITIVE_DECAL = 64,
+};
+
+class Shadow
+{
+
+public:
+    Shadow() : m_opacity(0xFF), m_color1(0xFFFFFFFF), m_color2(0xFFFFFFFF), m_angle(0.0f) {}
```
Comment (tomsons26) [inline src/game/client/shadow/w3dshadow.h]: init remaining members comment as bugfix

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 14]
```diff
@@ -0,0 +1,1581 @@
...
+            objectToLight.Z = 0.0f;
+            objectToLight.Normalize();
+        } else {
+            objectToLight.Set(1.0f, 0.0f, 0.0f);
+        }
+
+        struct SurfaceClass::SurfaceDescription surface_desc;
+        m_shadowTexture[0]->Get_Texture()->Get_Level_Description(surface_desc, 0);
+        Vector3 uVec = (objectToLight * 3.2f) / surface_desc.width;
+        Vector3 vVec = (objectToLight * 3.2f) / surface_desc.width;
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: * 3.2f) / surface_desc.height;
 also missing a Rotate_Z call here between the vectors

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 15]
```diff
@@ -0,0 +1,1581 @@
...
+            objectToLight = light_pos - objPos;
+            objectToLight.Z = 0.0f;
+            objectToLight.Normalize();
+        } else {
+            objectToLight.Set(1.0f, 0.0f, 0.0f);
+        }
+
+        struct SurfaceClass::SurfaceDescription surface_desc;
+        m_shadowTexture[0]->Get_Texture()->Get_Level_Description(surface_desc, 0);
+        Vector3 uVec = (objectToLight * 3.2f) / surface_desc.width;
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: surface members probably need casts?

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 16]
```diff
@@ -0,0 +1,1581 @@
...
+        if (objPos == Vector3(0.0f, 0.0f, 0.0f)) {
+            return;
+        }
+
+        Vector3 objToLight = light_pos - objPos;
+        objToLight.Normalize();
+        objToLight = objPos + (objToLight * 2000.0f);
+        m_shadowProjector->Compute_Perspective_Projection(m_robj, objToLight, -1.0f, -1.0f);
+        m_shadowProjector->Set_Render_Target(g_theW3DProjectedShadowManager->Get_Render_Target(), nullptr);
+        SpecialRenderInfoClass *rinfo = g_theW3DProjectedShadowManager->Get_Shadow_Context();
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: new line

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 17]
```diff
@@ -0,0 +1,1581 @@
...
+        rinfo->m_lightEnvironment->Reset(m_robj->Get_Position(), Vector3(0.0f, 0.0f, 0.0f));
+        m_shadowProjector->Compute_Texture(m_robj, rinfo);
+        SurfaceClass *src = m_shadowTexture[0]->Get_Texture()->Get_Surface_Level(0);
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp:319]: new line

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 18]
```diff
@@ -0,0 +1,1581 @@
...
+        SurfaceClass *dst = g_theW3DProjectedShadowManager->Get_Render_Target()->Get_Surface_Level(0);
+        src->Copy(0, 0, 0, 0, 512, 512, dst);
+        Ref_Ptr_Release(dst);
+        Ref_Ptr_Release(src);
+        m_shadowTexture[0]->Update_Bounds(g_theW3DShadowManager->Get_Light_Pos_World(0), m_robj);
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp:326]: new line

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 19]
```diff
@@ -0,0 +1,1581 @@
...
+        } else {
+            objectToLight.Set(1.0f, 0.0f, 0.0f);
+        }
+
+        struct SurfaceClass::SurfaceDescription surface_desc;
+        m_shadowTexture[0]->Get_Texture()->Get_Level_Description(surface_desc, 0);
+        Vector3 uVec = (objectToLight * 3.2f) / surface_desc.width;
+        Vector3 vVec = (objectToLight * 3.2f) / surface_desc.width;
+        m_shadowTexture[0]->Set_Decal_UV_Axis(uVec, vVec);
+        AABoxClass box;
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp:346]: new line

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 20]
```diff
@@ -0,0 +1,1581 @@
...
+    shadow->m_decalOffsetU = 1.0f / x;
+    shadow->m_decalOffsetV = 1.0f / y;
+    shadow->m_offsetX = offsetx;
+    shadow->m_offsetY = offsety;
+    shadow->m_flags = flags;
+    shadow->Init();
+    W3DProjectedShadow *s = nullptr;
+    W3DProjectedShadow *i;
+
+    for (i = m_decalList; i; i = i->m_next) {
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: needs to use m_shadowList

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 21]
```diff
@@ -0,0 +1,1581 @@
...
+        if (i->m_shadowTexture[0] = tex) {
+            shadow->m_next = i;
+
+            if (s) {
+                s->m_next = shadow;
+            } else {
+                m_decalList = shadow;
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: needs to use m_shadowList

File: src/game/client/shadow/w3dprojectedshadow.cpp [hunk 22]
```diff
@@ -0,0 +1,1581 @@
...
+
+            break;
+        }
+
+        s = i;
+    }
+
+    if (!i) {
+        shadow->m_next = m_decalList;
+        m_decalList = shadow;
```
Comment (tomsons26) [inline src/game/client/shadow/w3dprojectedshadow.cpp]: needs to use m_shadowList

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp:267]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp:408]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp:492]: It takes an FVF so 0 is appropriate here (or one or more of the D3DFVF constants)

File: src/game/client/shadow/w3dprojectedshadow.h (hunk 6 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.h]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp:319]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp:346]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp:326]: fixed.

File: src/game/client/shadow/w3dshadow.h (hunk 13 above)
Comment (jonwil) [reply src/game/client/shadow/w3dshadow.h]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 22 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

File: src/game/client/shadow/w3dprojectedshadow.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/client/shadow/w3dprojectedshadow.cpp]: fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**